File: 2024/9/dll.go
//go:build go_dll

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Node struct {
	Pos, Idx, Size int
	Prev, Next     *Node
}

func (n *Node) End() int {
	return n.Pos + n.Size
}

func (n *Node) InsertAfter(node *Node) *Node {
	if n.Next != nil {
		n.Next.Prev = node
		node.Next = n.Next
	}
	n.Next = node
	node.Prev = n
	return n.Next
}

func (n *Node) InsertBefore(node *Node) *Node {
	if n.Prev != nil {
		n.Prev.Next = node
		node.Prev = n.Prev
	}
	n.Prev = node
	node.Next = n
	return n.Prev
}

func (n *Node) Remove() {
	if n.Prev != nil {
		n.Prev.Next = n.Next
	}
	if n.Next != nil {
		n.Next.Prev = n.Prev
	}
	n.Prev, n.Next = nil, nil
}

type Disk struct {
	GapHead    *Node
	Blocks     *Node
	BlocksTail *Node
	Index      map[int]*Node
}

func NewDisk(data []int) *Disk {
	d := &Disk{Index: make(map[int]*Node)}
	pos := 0

	for i := 0; i < len(data); i++ {
		if i%2 == 0 {
			bNode := &Node{Pos: pos, Idx: i / 2, Size: data[i]}
			d.Index[i/2] = bNode
			if d.Blocks == nil {
				d.GapHead = bNode
				d.Blocks = bNode
				d.BlocksTail = bNode
			} else {
				d.BlocksTail = d.BlocksTail.InsertAfter(bNode)
			}
			pos += data[i]
		} else {
			pos += data[i]
		}
	}
	return d
}

func (d *Disk) LeftmostGap(size int) *Node {
	if d.GapHead == nil || d.GapHead.Next == nil {
		return d.GapHead
	}
	for d.GapHead.Next != nil && d.GapHead.End() == d.GapHead.Next.Pos {
		d.GapHead = d.GapHead.Next
	}

	res := d.GapHead
	for res.Next != nil && res.Next.Pos-res.End() < size {
		res = res.Next
		if res.Next == nil {
			return nil
		}
	}
	return res
}

func (d *Disk) MoveLeft(idx int, node *Node, count int) {
	idxNode := d.Index[idx]
	if node.Pos >= idxNode.End() {
		return
	}
	idxNode.Size -= count
	if node.Idx == idx {
		node.Size += count
	} else {
		node.InsertAfter(&Node{Pos: node.End(), Idx: idx, Size: count})
	}
}

func (d *Disk) RevFiles() []struct{ Idx, Size int } {
	var result []struct{ Idx, Size int }
	for idx := range d.Index {
		result = append(result, struct{ Idx, Size int }{Idx: idx, Size: d.Index[idx].Size})
	}
	// Sort in descending order of index
	for i := 0; i < len(result)-1; i++ {
		for j := i + 1; j < len(result); j++ {
			if result[i].Idx < result[j].Idx {
				result[i], result[j] = result[j], result[i]
			}
		}
	}
	return result
}

func (d *Disk) String() string {
	var res []rune
	node := d.Blocks
	pos := 0
	for node != nil {
		for i := 0; i < node.Pos-pos; i++ {
			res = append(res, '.')
		}
		for i := 0; i < node.Size; i++ {
			res = append(res, rune('0'+node.Idx))
		}
		pos = node.End()
		node = node.Next
	}
	return string(res)
}

func (d *Disk) Checksum() int {
	check := 0
	node := d.Blocks
	for node != nil {
		check += (((2*node.Pos + node.Size - 1) * node.Size) / 2) * node.Idx
		node = node.Next
	}
	return check
}

func sol1(data []int) int {
	d := NewDisk(data)
	for _, file := range d.RevFiles() {
		idx, size := file.Idx, file.Size
		for i := 0; i < size; i++ {
			gap := d.LeftmostGap(1)
			if gap == nil {
				continue
			}
			d.MoveLeft(idx, gap, 1)
		}
	}
	return d.Checksum()
}

func sol2(data []int) int {
	d := NewDisk(data)
	for _, file := range d.RevFiles() {
		idx, size := file.Idx, file.Size
		gap := d.LeftmostGap(size)
		if gap == nil {
			continue
		}
		d.MoveLeft(idx, gap, size)
	}
	return d.Checksum()
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	if scanner.Scan() {
		line := scanner.Text()
		data := make([]int, len(line))
		for i, c := range line {
			data[i], _ = strconv.Atoi(string(c))
		}
		fmt.Println(sol1(data))
		fmt.Println(sol2(data))
	}
}

File: 2024/9/dll.py
import sys
import dataclasses
import tqdm


@dataclasses.dataclass
class Node:
    pos: int
    idx: int
    size: int
    prev: "Node" = None
    next: "Node" = None

    def end(self):
        return self.pos + self.size

    def insert_after(self, node):
        if self.next is not None:
            self.next.prev = node
            node.next = self.next
        self.next = node
        self.next.prev = self
        return self.next

    def insert_before(self, node):
        if self.prev is not None:
            self.prev.next = node
            node.prev = self.prev
        self.prev = node
        self.prev.next = self
        return self.prev

    def remove(self):
        self.prev.next = self.next
        self.next.prev = self.prev
        self.prev = self.next = None


class Disk:
    def __init__(self, data):
        self.gap_head = None
        self.blocks = None
        self.blocks_tail = None
        self.index = {}
        pos = 0
        for i in range(len(data)):
            if i % 2 == 0:
                bnode = Node(pos, i // 2, data[i])
                self.index[i // 2] = bnode
                if self.blocks is None:
                    self.gap_head = bnode
                    self.blocks = bnode
                    self.blocks_tail = bnode
                else:
                    self.blocks_tail = self.blocks_tail.insert_after(bnode)
                pos += data[i]
            else:
                pos += data[i]

    def leftmost_gap(self, size=1):
        if self.gap_head is None or self.gap_head.next is None:
            return self.gap_head
        while (
            self.gap_head.next is not None
            and self.gap_head.end() == self.gap_head.next.pos
        ):
            self.gap_head = self.gap_head.next

        res = self.gap_head
        while res.next.pos - res.end() < size:
            res = res.next
            if res.next is None:
                return None
        return res

    def move_left(self, idx, node, count=1):
        idx_n = self.index[idx]
        if node.pos >= idx_n.end():
            return
        idx_n.size -= count
        if node.idx == idx:
            node.size += count
        else:
            node.insert_after(Node(node.end(), idx, count))

    def rev_files(self):
        idxs = list(self.index.keys())
        idxs.sort(reverse=True)
        return [(idx, self.index[idx].size) for idx in idxs]

    def __repr__(self):
        return str(self)

    def __str__(self):
        res = []
        node = self.blocks
        pos = 0
        while node is not None:
            res.extend(["."] * (node.pos - pos))
            res.extend([str(node.idx)] * node.size)
            pos = node.end()
            node = node.next
        return "".join(res)

    def checksum(self):
        check = 0
        node = self.blocks
        while node is not None:
            check += (((2 * node.pos + node.size - 1) * node.size) // 2) * node.idx
            node = node.next
        return check


def sol1(data):
    D = Disk(data)
    for idx, size in D.rev_files():
        for _ in range(size):
            g = D.leftmost_gap()
            if g is None:
                continue
            D.move_left(idx, g)
    return D.checksum()


def sol2(data):
    D = Disk(data)
    for idx, size in D.rev_files():
        g = D.leftmost_gap(size)
        if g is None:
            continue
        D.move_left(idx, g, size)
    return D.checksum()


if __name__ == "__main__":
    data = [int(x) for x in sys.stdin.readline().strip()]
    print(sol1(data))
    print(sol2(data))
